Cleanup internal ComWrappers cache when object enters Finalization queue

AaronRobinsonMSFT opened the pull request:
The `ComWrappers` API has a deficiency/race during native object wrapper (RCW) cleanup. This is rare but has been observed in tight loops with complex object types. The scenario occurs when native object A is freed, its memory is reused for native object B, and that memory is provided back to the runtime but the previous mapping remains to the wrapper for native object A. The issue occurs based on an incorrect expectation between when the RCW's Finalizer will run and when the mapping between native `IUnknown*` to RCW will be removed from the internal cache.

The solution is to remove the mapping when the RCW itself is scheduled for Finalization. 

The issue was reported at https://github.com/microsoft/CsWinRT/issues/762.

Additional issues
Fixes https://github.com/dotnet/runtime/issues/53160

AaronRobinsonMSFT committed:
Clean up the ExternalObjectContext cache for ComWrappers' RCWs.

AaronRobinsonMSFT committed:
Add testing for verifying internal cache clean up.

AaronRobinsonMSFT committed:
Remove commented out tests - oops.

AaronRobinsonMSFT commented:
/cc @elinor-fung @jkoritzinsky

AaronRobinsonMSFT commented:
Failure is https://github.com/dotnet/runtime/issues/52710

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into cleanup_cache

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into cleanup_cache

AaronRobinsonMSFT committed:
Fix case for Disconnect logic.

AaronRobinsonMSFT commented on src/coreclr/vm/interoplibinterface.cpp:
```diff
@@ -30,9 +30,15 @@ namespace
             Flags_InCache = 4,
+
+            // The EOC is "detached" and no longer used to map between identity and a managed object.
+            Flags_Detached = 8,
```
Comment on detach/collected relationship.

AaronRobinsonMSFT commented on src/coreclr/vm/interoplibinterface.h:
```diff
@@ -187,7 +188,7 @@ class Interop
     // Notify before/after when GC is scanning roots.
     // Present assumption is that calls will never be nested.
     static void OnBeforeGCScanRoots();
-    static void OnAfterGCScanRoots();
```
When do these GC callbacks occur - let's doc this more clearly.

AaronRobinsonMSFT commented on src/coreclr/vm/interoplibinterface_shared.cpp:
```diff
@@ -139,7 +139,12 @@ void Interop::OnAfterGCScanRoots()
     }
     CONTRACTL_END;
 
+#ifdef FEATURE_COMWRAPPERS
```
Why is one okay during concurrent and the other not?

AaronRobinsonMSFT committed:
Review feedback.

elinor-fung commented on src/tests/Interop/COM/ComWrappers/API/Program.cs:
```diff
@@ -280,6 +280,56 @@ static void ValidatePrecreatedExternalWrapper()
+                nativeWrapper = (ITrackerObjectWrapper)cw.GetOrRegisterObjectForComInstance(trackerObjRaw, CreateObjectFlags.None, nativeWrapper);
+
+                // Set this on the return instead of during creation since the returned wrapper may be the one from
+                // the internal cache and not the one passed in above.
```
From the usage, aren't we intentionally making it so that the first wrapper should have been removed from the cache when we create/register the second wrapper?

AaronRobinsonMSFT replied:
Yes. This is creating the function in an agnostic way since the returned value could be different from the input.

elinor-fung replied:
I get that. I just find it confusing to have this local function - specifically for creating and registering a wrapper - explicitly say that the created and registered wrapper could be different when it should never be based on the usage and that fact is actually a key part of this test.

We can agree to disagree.

AaronRobinsonMSFT commented on src/coreclr/vm/interoplibinterface.h:
```diff
@@ -187,7 +188,7 @@ class Interop
     // Present assumption is that calls will never be nested.
     static void OnBeforeGCScanRoots();
-    static void OnAfterGCScanRoots();
+    static void OnAfterGCScanRoots(_In_ bool is_concurrent);
```
Just to match the GC. Do you want me to change it?

AaronRobinsonMSFT committed:
Review feedback.

AaronRobinsonMSFT committed:
Detaching only works when inserted into the cache. Since not inserting
is an option remove the use of Detach in the IsActive logic and
narrow its use to the cache cleanup scenario only.
